src/Common/GuardedPoolAllocator.h

```diff
@@ -0,0 +1,111 @@
+
+    inline ALWAYS_INLINE bool shouldSample()
+    {
+        // next_sample_counter == 0 means we "should regenerate the counter".
+        //                   == 1 means we "should sample this allocation".
+        // AdjustedSampleRatePlusOne is designed to intentionally underflow. This
+        // class must be valid when zero-initialised, and we wish to sample as
+        // infrequently as possible when this is the case, hence we underflow to
+        // UINT32_MAX.
+        if (unlikely(getThreadLocals()->next_sample_counter == 0))
+            getThreadLocals()->next_sample_counter = (getRandomUint32T() % (adjusted_sample_rate_plus_one - 1)) + 1;
+        return unlikely(--getThreadLocals()->next_sample_counter == 0);
+    }
+
+    inline ALWAYS_INLINE bool pointerIsMine(const void * ptr) const { return state.pointerIsMine(ptr); }
+
+    // Returns a pointer to the Metadata region, or nullptr if it doesn't exist.
+    const AllocationMetadata * getMetadataRegion() const { return metadata; }
+
+    // Returns a pointer to the AllocatorState region.
+    const GuardedPoolAllocatorState * getAllocatorState() const { return &state; }
+
+private:
+    static constexpr size_t kInvalidSlotID = SIZE_MAX;
+
+    /// Map and unmap memory with mmap
+    static void * map(size_t Size);
+
+    static void * reserveGuardedPool(size_t size);
+
+    size_t reserveSlotInGuardedPool();
+    // Unreserve the guarded slot.
+    void freeSlotInGuardedPool(size_t slot_index);
+
+    // allocateInGuardedPool() ptr and size must be a subrange of the previously
+    // reserved pool range.
+    void allocateInGuardedPool(void * ptr, size_t size) const;
+    // deallocateInGuardedPool() ptr and size must be an exact pair previously
+    // passed to allocateInGuardedPool().
+    void deallocateInGuardedPool(void * ptr, size_t size) const;
+
+    static uintptr_t alignUp(uintptr_t ptr, size_t alignment);
+    static uintptr_t alignDown(uintptr_t ptr, size_t alignment);
+
+    AllocationMetadata * addrToMetadata(uintptr_t ptr) const;
+
+    [[noreturn]] void trapOnAddress(uintptr_t address, Error err);
+
+    static size_t getPlatformPageSize();
+
+    /// Return random uint32_t from [0, UINT32_MAX] with
+    /// std::uniform_int_distribution<uint32_t>
+    uint32_t getRandomUint32T() const;
+
+    class ScopedRecursiveGuard
+    {
+    public:
+        ScopedRecursiveGuard() { getThreadLocals()->recursive_guard = true; }
+        ~ScopedRecursiveGuard() { getThreadLocals()->recursive_guard = false; }
+    };
+
+    /// Array of free slots and its size. When a slot is freed it is added
+    /// to this array. When a free slot is requested, a random element of this
+    /// array is chosen, swapped with last element and the array is shrunk.
+    size_t * free_slots = nullptr;
+    size_t free_slots_length = 0;
+
+    size_t adjusted_sample_rate_plus_one = 0;
+
+    size_t currently_allocated_slots = 0;
+
+    AllocationMetadata * metadata;
```

Comment:
Is it required here? I understand that technically it is a memory leak, but we need this memory for as long as the program runs (and GWP is one of the last destroyed variables along with other static variables).
And I guess we have to also unmap pool itself and free slots if we choose to unmap metadata.

Reply:
To make it leak-free yes (note, that right now LSan is broke in clickhouse, see #33957), but it is not that important, and besides it may be tricky to free it once nobody uses it, so you can ignore this.